backend/src/friend/friend.service.ts

```diff
@@ -1,4 +1,452 @@
+    const checkData = await this.checkSenderAndReceiver(user, payload);
+    if (checkData !== '') {
+      throw new BadRequestException(checkData);
+    }
+
+    const senderEmail: string = user.email;
+    const receiverEmail: string = payload.userEmail;
+
+    // find the receiver in database
+    const sender: User = user;
+    const receiver: User = await this.getUserByEmail(receiverEmail);
+    if (!receiver) {
+      throw new BadRequestException('Receiver not found.');
+    }
+
+    // check if receiver's receivedRequest array contains the sender's email
+    if (receiver.friendRequestsReceived.includes(senderEmail)) {
+      throw new BadRequestException(
+        "Friend request already in receiver's list",
+      );
+    }
+
+    // check if sender's sentRequest array contains the receiver's email
+    if (sender.friendRequestsSent.includes(receiverEmail)) {
+      throw new BadRequestException("Friend request already in sender's list.");
+    }
+
+    // check if receiver was the sender
+    if (
+      receiver.friendRequestsSent.includes(senderEmail) ||
+      sender.friendRequestsReceived.includes(receiverEmail)
+    ) {
+      throw new BadRequestException('The receiver sent you an request.');
+    }
+
+    // check if they are already friends
+    if (
+      sender.friends.includes(receiverEmail) ||
+      receiver.friends.includes(senderEmail)
+    ) {
+      throw new BadRequestException('You are already friends.');
+    }
+
+    // update receiver's list
+    try {
+      const updateReceiver = await this.prisma.user.update({
+        where: {
+          email: receiverEmail,
+        },
+        data: {
+          friendRequestsReceived: {
+            push: senderEmail,
+          },
+        },
+      });
+    } catch (error) {
+      throw new ServiceUnavailableException('Server unavailable.');
+    }
+
+    // update sender's list
+    try {
+      const updateReceiver = await this.prisma.user.update({
+        where: {
+          email: senderEmail,
+        },
+        data: {
+          friendRequestsSent: {
+            push: receiverEmail,
+          },
+        },
+      });
+    } catch (error) {
+      throw new ServiceUnavailableException('Server unavailable.');
+    }
+
+    return {
+      status: 'OK',
+      statusCode: 200,
+      message: 'Friend request successufly sent.',
+    };
+  }
+
+  /*****************************************************************************/
+  /* cancel friend Request */
+  /*****************************************************************************/
+  async cancelFriendRequest(
+    user: User,
+    payload: { userEmail: string },
+    res: Response,
+  ) {
+    // check user/payload data
+    const checkData = await this.checkSenderAndReceiver(user, payload);
+    if (checkData !== '') {
+      throw new BadRequestException(checkData);
+    }
+
+    const senderEmail: string = user.email;
+    const receiverEmail: string = payload.userEmail;
```

Comment:
These two lines can be added to the checkSenderAndReceiver function to avoid repetitive code, and to include error handling.